transformer_engine/jax/cpp_extensions.py

```diff
@@ -209,1824 +230,2350 @@ def custom_caller(name, args, opaque, has_side_effect, **kwargs):
-        out_dtype = te_dtype_to_jax_dtype(dtype)
+        x_dtype = dtypes.canonicalize_dtype(x_aval.dtype)
+        assert x_dtype in [jnp.float32, jnp.float16, jnp.bfloat16]
 
-        assert len(inputs.shape) == 2
-        assert isinstance(dtype, TEDType)
-        assert in_dtype == out_dtype
+        mu_rsigama_dtype = jnp.float32
 
-        return ShapedArray((inputs.shape[1], inputs.shape[0]),
-                           in_dtype,
-                           named_shape=inputs.named_shape)
+        out_aval = core.raise_to_shaped(x_aval)
+        mu_aval = rsigma_aval = out_aval.update(shape=out_aval.shape[:-1], dtype=mu_rsigama_dtype)
+
+        assert gamma_aval.size == beta_aval.size
+        hidden_size = gamma_aval.size
+        assert x_aval.size % hidden_size == 0
+
+        return out_aval, mu_aval, rsigma_aval
 
     @staticmethod
-    def lowering(ctx, inputs, *, dtype):
+    def lowering(ctx, x, gamma, beta, *, zero_centered_gamma, epsilon):
         """
-        _transpose cuda lowering
+        LayerNorm fwd lowering rules
         """
+        x_aval, gamma_aval, beta_aval = ctx.avals_in
+        assert gamma_aval.dtype == beta_aval.dtype
+        x_type = ir.RankedTensorType(x.type)
+        x_shape = x_type.shape
+        w_type = ir.RankedTensorType(gamma.type)
+        w_shape = w_type.shape
+        b_type = ir.RankedTensorType(beta.type)
+        b_shape = b_type.shape
 
-        in_aval = ctx.avals_in[0]
-        assert in_aval.dtype in [jnp.float32, jnp.float16, jnp.bfloat16, jnp.int8]
+        assert w_type == b_type
+        assert w_shape == b_shape
 
-        ir_in_type = ir.RankedTensorType(inputs.type)
-        ir_in_shape = ir_in_type.shape
-        ir_out_dtype = te_dtype_to_ir_dtype(dtype)
+        # Output shape is same as the input shape, but the output type is same as the weight type.
+        # See ln_api.cpp
+        output_type = w_type.element_type
+        ir_mu_dtype = ir.F32Type.get()
+        ir_rsigma_dtype = ir.F32Type.get()
 
-        out_types = [ir.RankedTensorType.get([ir_in_shape[1], ir_in_shape[0]], ir_out_dtype)]
-        operands = [inputs]
-        operand_shapes = [ir_in_shape]
+        out_shape = x_shape
+        hidden_size = reduce(operator.mul, w_shape)
+        batch_shape = out_shape[:-1]
+        batch_size = reduce(operator.mul, x_shape) // hidden_size
+
+        out_types = [
+            ir.RankedTensorType.get(out_shape, output_type),
+            ir.RankedTensorType.get(batch_shape, ir_mu_dtype),
+            ir.RankedTensorType.get(batch_shape, ir_rsigma_dtype),
+        ]
+        operands = [x, gamma, beta]
+        operand_shapes = [x_shape, w_shape, b_shape]
         args = CustomCallArgsWrapper(out_types, operands, operand_shapes)
 
-        assert len(ir_in_shape) == 2
-        opaque = transformer_engine_jax.pack_common_descriptor(ir_in_shape, dtype, dtype)
+        opaque = transformer_engine_jax.pack_norm_descriptor(
```

Comment:
I didn't see any check about the input tensor rank. Is that normal?

Reply:
Could you explain a bit about checking the input tensor rank.

Reply:
The ops support multi-dimensions now, so there is no need to check the rank for inputs.